pytorch_lightning/core/lightning.py

```diff
@@ -1568,13 +1574,24 @@ def to_torchscript(
 
         mode = self.training
         with torch.no_grad():
-            scripted_module = torch.jit.script(self.eval(), **kwargs)
+            if method == 'script':
+                torchscript_module = torch.jit.script(self.eval(), **kwargs)
+            elif method == 'trace':
+                # if no example inputs are provided, try to see if model has example_input_array set
+                if example_inputs is None:
+                    example_inputs = self.example_input_array
+                # automatically send example inputs to the right device and use trace
+                torchscript_module = torch.jit.trace(func=self.eval(), example_inputs=example_inputs.to(self.device),
```

Comment:
Here you assume that `example_input_array` is a tensor, but this is not true. 
If forward takes `*args`, `example_input_array` is a tuple, and if forward takes `**kwargs`, `example_input_array` must be a dict.

Reply:
Given your comment on the issue that `.trace()` accepts either a `tuple` or a `torch.Tensor` (that is automatically converted to a tuple), it means that the input should be: `example_input_array: Optional[Union[torch.Tensor, Tuple[torch.Tensor]]]`?

However, when the forward function accepts `**kwargs`, `self.example_input_array` could be a dict, in which case `.trace(example_inputs=example_inputs)` will fail?

What would be the best way to approach this? Does this mean that `.trace()` cannot be used if `forward` expects a dict?

Reply:
```python 
import torch
import torch.nn as nn

class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(1, 1, 3)

    def forward(self, x):
        return self.conv(x)


class Net2(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(1, 1, 3)

    def forward(self, x, y):
        return self.conv(x)


class Net3(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(1, 1, 3)

    def forward(self, x, y):
        return self.conv(x)

# SINGLE INPUT
net = Net()
ex_inp = torch.rand(1, 1, 3, 3)
torch.jit.trace(net, ex_inp)

# TWO INPUTS
net = Net2()
torch.jit.trace(net, (ex_inp, ex_inp))

# DICT (**kwargs)
# fails
# net = Net3()
# torch.jit.trace(net, dict(x=ex_inp, y=ex_inp))
```

Here is an example. tracing supports single input and tuple, which gets unrolled to multiple positional args. In these two cases, you can use the Lightning self.example_input_array. However, dicts will not be passed as kwargs, and instead as a single input. In Lightning however, a dict would mean **kwargs. 

Two ways to solve this: 
- leave as is, user needs to know how self.example_input_array works
- error when `self.example_input_array` is a dict
- do not even use `self.example_input_array`, and require the user to give inputs to the method directly

Then there is a second issue. You should use the `pytorch_lightning.utilities.apply_func.move_data_to_device` to move the example input to the device, since it could be a tuple.

Reply:
cc @ananthsub

Reply:
1 & 2 could be combined by raising a warning instead of an error. From PL's side throw a warning similar to:

> `self.example_input_array` cannot be a dict. Please provide a sample Tensor/Tuple to `example_inputs` as argument, or set `self.example_input_array` to a Tensor/Tuple.

Then output the actual error produced by `.trace()`.
If in the future `.trace()` would be updated to support a dict, there is no need for a change (except removing the warning) on PL's side.

Personally, PL is for me about removing boilerplate code. Since `self.example_input_array` is already a thing in PL, it's better to use it. Therefore, I would advise against option 3.
I haven't used `self.example_input_array` personally yet, but in how many projects would this be a dict?

Reply:
yes, makes sense. 
Would you like to follow up on this with a PR? Would greatly appreciate this. For me the main concern is to properly move the input to the device with the function I referenced. For the way inputs are passed in, I don't have a strong opinon.

Reply:
IDK how much future support there will be for tracing vs scripting (scripting is strongly recommended). Rather than adding more trace support at the top-level of the PL module, why not override `to_torchscript` in your lightning module to determine how you want to export? then you have way more flexibility with tracing

Reply:
@awaelchli Ok, I'll follow up with another pull request using the `move_data_to_device` function.

@ananthsub Could you elaborate on why scripting is strongly recommended? From the [TorchScript documentation](https://pytorch.org/docs/stable/jit.html), scripting and tracing just focuses on different use cases, without a preference for either one. Scripting is positioned as useful when you need control-flow, while trace is simpler to use (no code changes).
I would go into more detail, but I think this discussion is more suitable under the original request: https://github.com/PyTorchLightning/pytorch-lightning/issues/4140

Reply:
@awaelchli I addressed your issues in a follow-up pull request (could not be added to this one due to it already being merged):
https://github.com/PyTorchLightning/pytorch-lightning/pull/4360